src/Security/EventListener/DenyAccessListener.php

```diff
@@ -81,7 +81,17 @@ public function onKernelRequest(GetResponseEvent $event)
         $extraVariables['request'] = $request;
 
         if (!$this->resourceAccessChecker->isGranted($attributes['resource_class'], $isGranted, $extraVariables)) {
-            throw new AccessDeniedException();
+            $defaultMessage = 'Access Denied.';
+
+            if (isset($attributes['collection_operation_name'])) {
+                $message = $resourceMetadata->getCollectionOperationAttribute($attributes['collection_operation_name'], 'access_control_message', $defaultMessage, true);
+            } elseif (isset($attributes['item_operation_name'])) {
+                $message = $resourceMetadata->getItemOperationAttribute($attributes['item_operation_name'], 'access_control_message', $defaultMessage, true);
+            } else {
+                $message = $resourceMetadata->getCollectionOperationAttribute($attributes['subresource_operation_name'], 'access_control_message', $defaultMessage, true);
+            }
```

Comment:
Couldn't we add a `getOperationAttribute` method that does the if/elseif/else thing in `resourceMetadata` to avoid all this boilerplate?

Reply:
Good idea. But `getOperationAttribute` already exists as a private method :) but this would be out of PR scope, isn't it?

Reply:
yeah it would and anyway we have a lot of these in the current code I'm going to open an issue on that matter.